### PR TITLE
[Examples] Add support for Min-SNR weighting strategy for better convergence

### PR DESCRIPTION
**Pull request by sayakpaul**
Paper reference: https://arxiv.org/abs/2303.09556
Code: https://github.com/TiankaiHang/Min-SNR-Diffusion-Training

### CODE REVIEW

**Comment by sayakpaul** [comment]
> @patil-suraj can you take a look here?

Actually, I would suggest not to. I am still gathering evidence to see if the PR is worth merging or even reviewing. After I am done, I will update here. Till then, don't worry about it.

File: examples/text_to_image/train_text_to_image.py [hunk 1]
```diff
@@ -325,9 +362,83 @@ def get_full_repo_name(model_id: str, organization: Optional[str] = None, token:
...
+    def fn(timesteps):
+        alpha = expand_tensor(sqrt_alphas_cumprod, timesteps, timesteps.shape)
+        sigma = expand_tensor(sqrt_one_minus_alphas_cumprod, timesteps, timesteps.shape)
+        snr = (alpha / sigma) ** 2
+        return snr
+
+    return fn
+
+
+def log_validation(vae, text_encoder, tokenizer, unet, args, accelerator, weight_dtype, epoch):
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image.py:393]: is this related to this PR title or does it just add logging?

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -476,6 +589,9 @@ def load_model_hook(models, input_dir):
             args.learning_rate * args.gradient_accumulation_steps * args.train_batch_size * accelerator.num_processes
         )
 
+    if args.snr_gamma is not None:
+        snr_fn = compute_snr(noise_scheduler)
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image.py:593]: This approach is too functional for me. I'm not a fan of having functions return other functions that's difficult to debug. Can we just write down a single function `compute_snr_loss(...)` that is called everytime below?

**Comment by patrickvonplaten** [comment]
Would be great to stick to a more imperative coding style which we prefer in libraries like `diffusers` and `transformers` (https://stackoverflow.com/questions/21895525/python-programming-functional-vs-imperative-code) 

I don't think it's very easy to follow if a function returns another function etc... this is a bit too jaxy/TF-like to me

File: examples/text_to_image/train_text_to_image.py (hunk 1 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:393]: The PR could have been without this utility but it was important to have it in the PR because otherwise, it was difficult to validate the effectiveness of the method. 

FWIW, I am not a fan of adding unrelated changes in a PR but this seemed important.

**Comment by sayakpaul** [comment]
@patrickvonplaten the latest changes should have addressed all your concerns. PTAL.

File: examples/text_to_image/train_text_to_image.py [hunk 3]
```diff
@@ -457,6 +564,9 @@ def load_model_hook(models, input_dir):
             args.learning_rate * args.gradient_accumulation_steps * args.train_batch_size * accelerator.num_processes
         )
 
+    # if args.snr_gamma is not None:
+    #     snr_fn = compute_snr(noise_scheduler)
+
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:569]: should be deleted

File: examples/text_to_image/train_text_to_image.py [hunk 4]
```diff
@@ -715,7 +827,23 @@ def collate_fn(examples):
...
+                    snr = compute_snr(timesteps)
+                    mse_loss_weights = (
+                        torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(dim=1)[0] / snr
+                    )
+                    # We first calculate the original loss. Then we mean over the non-batch dimensions and
+                    # rebalance the sample-wise losses with their respective loss weights.
+                    # Finally, we take the mean of the rebalanced loss.
+                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="none")
+                    loss = loss.mean(dim=list(range(1, len(loss.shape)))) * mse_loss_weights
+                    loss = loss.mean()
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:846]: Nice!

File: examples/text_to_image/train_text_to_image.py [hunk 5]
```diff
@@ -715,7 +827,23 @@ def collate_fn(examples):
 
                 # Predict the noise residual and compute loss
                 model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
-                loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")
+
+                if args.snr_gamma is None:
+                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")
+                else:
+                    # Compute loss-weights as per Section 3.4 of https://arxiv.org/abs/2303.09556.
+                    # Since we predict the noise instead of x_0, the original formulation is slightly changed.
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:835]: some models (sd2.1 and above) use `v-prediction`, does this formulation also work with v-prediction?

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding support for this! The loss curve does look smoother than without using `min-snr-weighting`. Just left a small comment about `v-prediction`.

It would also be cool to add this `train_unconditional.py`, it would be easy to verify this there since we train from scratch.

File: examples/text_to_image/train_text_to_image.py (hunk 5 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:835]: Yes sir. 

https://github.com/TiankaiHang/Min-SNR-Diffusion-Training/blob/521b624bd70c67cee4bdf49225915f5945a872e3/guided_diffusion/gaussian_diffusion.py#L854

**Comment by sayakpaul** [comment]
@patil-suraj when you get a moment could you review the changes introduced in https://github.com/huggingface/diffusers/pull/2899/commits/96e725433b2bad026c1d5104bcbaa406907dbf72? All of them are related to documentation. 

I think then we can merge. @patrickvonplaten maybe you also want to take a look.

File: examples/text_to_image/train_text_to_image.py [hunk 6]
```diff
@@ -41,15 +41,74 @@
...
 check_min_version("0.15.0.dev0")
 
 logger = get_logger(__name__, log_level="INFO")
 
+DATASET_NAME_MAPPING = {
+    "lambdalabs/pokemon-blip-captions": ("image", "text"),
+}
+
+
+def log_validation(vae, text_encoder, tokenizer, unet, args, accelerator, weight_dtype, epoch):
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:62]: Added this method as a part of the PR as well. Handles EMA offload and unload properly to ensure inference is being done with the EMA'd checkpoints.

File: docs/source/en/training/text2image.mdx [hunk 7]
```diff
@@ -155,6 +155,28 @@ python train_text_to_image_flax.py \
 </jax>
 </frameworkcontent>
 
+## Training with better convergence
```
Comment (patil-suraj) [inline docs/source/en/training/text2image.mdx:158]: Maybe 
```suggestion
## Training with Min-SNR weighting
```

As I think we don't know (yet?) if it'll always lead to better convergence for text2image. But no strong preference :)

File: docs/source/en/training/text2image.mdx [hunk 8]
```diff
@@ -155,6 +155,28 @@ python train_text_to_image_flax.py \
...
+
+We support training with the Min-SNR weighting strategy proposed in [Efficient Diffusion Training via Min-SNR Weighting Strategy](https://arxiv.org/abs/2303.09556) which helps to achieve faster convergence
+by rebalancing the loss. In order to use it, one needs to set the `--snr_gamma` argument. The recommended
+value when using it is 5.0.
```
Comment (patil-suraj) [inline docs/source/en/training/text2image.mdx:162]: Out of curiosity, where is this value proposed? Is there a rule of thumb when choosing a value for this?

File: docs/source/en/training/text2image.mdx [hunk 9]
```diff
@@ -155,6 +155,28 @@ python train_text_to_image_flax.py \
...
+* Training with the Min-SNR weighting strategy (`snr_gamma` set to 5.0)
+* Training with the Min-SNR weighting strategy (`snr_gamma` set to 1.0)
+
+For our small Pokemons dataset, the effects of Min-SNR weighting strategy might not appear to be pronounced, but for larger datasets, we believe the effects will be more pronounced.
+
+Also, note that in this example, we either predict `epsilon` (i.e., the noise) or the `v_prediction`. For both of these cases, the formulation of the Min-SNR weighting strategy that we have used holds. 
+
+<Tip warning={true}>
+
+Training with Min-SNR weighting strategy is only supported in PyTorch.
```
Comment (patil-suraj) [inline docs/source/en/training/text2image.mdx:176]: for future PR: Could be cool to add this in jax as well, will be useful for the jax event.

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding the doc. Looks good!

File: docs/source/en/training/text2image.mdx (hunk 8 above)
Comment (sayakpaul) [reply docs/source/en/training/text2image.mdx:162]: It's reported in the paper. A gamma of 5.0 always leads to better results in the experiments presented by the authors in the paper.

File: examples/text_to_image/README.md [hunk 10]
```diff
@@ -111,6 +111,22 @@ image = pipe(prompt="yoda").images[0]
 image.save("yoda-pokemon.png")
 ```
 
+#### Training with better convergence
```
Comment (sayakpaul) [inline examples/text_to_image/README.md:114]:
```suggestion
#### Training with Min-SNR weighting
```

**Review by patrickvonplaten — APPROVED** [review]
Very cool! Thanks for iterating
